scipy/signal/_spectral_py.py

```diff
@@ -85,71 +111,200 @@ def lombscargle(x,
+        raise ZeroDivisionError("Freqs array contains a 0.")
+    
+    # check for weights summing to 0
+    if weights.sum() == 0:
+        raise ZeroDivisionError("Weights sum to 0.")
+    
+    # check for any negative weights
+    if np.any(weights < 0):
+        raise ValueError("Each weight must be >= 0.")
+
+    # validate normalize parameter
+    if isinstance(normalize, bool):
+        # if bool, convert to str literal
+        normalize = "normalize" if normalize else "power"
+    else:
+        # if neither a bool or str
+        if not isinstance(normalize, str):
+            raise TypeError("Normalize type must a bool or str.")
+
+    if normalize not in ["power", "normalize", "amplitude"]:
+        raise ValueError(
+            "Normalize must be: "
+            "False (or 'power'), "
+            "True (or 'normalize'), "
+            "or 'amplitude'."
+        )
+    
+    # for normalize == 'amplitude', check if freq_min < freq_critical
+    if (normalize == 'amplitude'):
+        freq_critical = 2*np.pi/(x.max() - x.min())  # rad/sec
+        if (freqs.min() < freq_critical):
+            message = (f"Frequencies below {freq_critical:.3f} rad/sec are longer than "
+                       "the provided time range, therefore their estimated amplitudes "
+                       "will tend towards infinity.")
+            warnings.warn(message, stacklevel=2)
+
+    # convert inputs to contiguous arrays with high precision
     x = np.ascontiguousarray(x, dtype=np.float64)
     y = np.ascontiguousarray(y, dtype=np.float64)
     freqs = np.ascontiguousarray(freqs, dtype=np.float64)
-
-    assert x.ndim == 1
-    assert y.ndim == 1
-    assert freqs.ndim == 1
-
-    if precenter:
-        pgram = _lombscargle(x, y - y.mean(), freqs)
-    else:
-        pgram = _lombscargle(x, y, freqs)
-
-    if normalize:
-        pgram *= 2 / np.dot(y, y)
-
+    weights = np.ascontiguousarray(weights, dtype=np.float64)
+
+    # weight vector must sum to 1
+    weights = weights / weights.sum()
+
+    # pre-allocate arrays for intermediate and final calculations
+    coswt = np.empty_like(x)
+    sinwt = np.empty_like(x)
+    wcoswt = np.empty_like(x)
+    wsinwt = np.empty_like(x)
+    pgram = np.empty_like(freqs)
+    # store a and b so that phase can be calculated outside loop, if necessary
+    a = np.empty_like(freqs)
+    b = np.empty_like(freqs)
+
+    # calculate the single sum that does not depend on the frequency
+    Y_sum = (weights * y).sum()
+
+    # loop over the frequencies
+    for i in range(freqs.shape[0]):
+        coswt[:] = np.cos(freqs[i] * x)
+        sinwt[:] = np.sin(freqs[i] * x)
+        wcoswt[:] = weights * coswt
+        wsinwt[:] = weights * sinwt
+
+        # these variables names are of form XX_hat in paper, but dropping the suffix 
+        # to reuse variables later whether floating_mean is True or False
+        YC = (y * wcoswt).sum()
+        YS = (y * wsinwt).sum()
+        CC = (wcoswt * coswt).sum()
```

Comment:
A potential micro-optimization for reducing memory use would be to move the calculation of `CC` and `CS` before the calculation of ``wcoswt`` and ``wsinwt``. Then ``coswt`` and ``sinwt`` could be eliminated by using ``wcoswt`` and ``wsinwt``. Those would be updated later by multiplying them by ``weights``.
Not sure if it has a measurable effect, though.

Reply:
I'll look into this. I know I had tested all sorts of ways to optimize on the original code. But I didn't reassess the optimizations after I added the `floating_mean` branch.

Reply:
Okay, I removed the 2 additional arrays. The code is a little more complex to follow, and reusing the variable names isn't ideal. But I think that ship sailed with making the `floating_mean` option.